apps/src/applab/applab.js

```diff
@@ -801,29 +807,181 @@ Applab.init = function(config) {
     if (codeModeButton) {
       dom.addClickTouchEvent(codeModeButton, Applab.onCodeModeButton);
     }
-    var designModeAddButton = document.getElementById('designModeAddButton');
-    if (designModeAddButton) {
-      dom.addClickTouchEvent(designModeAddButton, Applab.onDesignModeAddButton);
-    }
-    var designModeAddInput = document.getElementById('designModeAddInput');
-    if (designModeAddInput) {
-      dom.addClickTouchEvent(designModeAddInput, Applab.onDesignModeAddInput);
-    }
-    var designModeAddLabel = document.getElementById('designModeAddLabel');
-    if (designModeAddLabel) {
-      dom.addClickTouchEvent(designModeAddLabel, Applab.onDesignModeAddLabel);
-    }
     var designModeClear = document.getElementById('designModeClear');
     if (designModeClear) {
       dom.addClickTouchEvent(designModeClear, Applab.onDesignModeClear);
     }
 
+    // Allow elements to be dragged and dropped from the design mode
+    // element tray to the play space.
+    if (window.$) {
+      $('.new-design-element').draggable({
+        containment:"#codeApp",
+        helper:"clone",
+        appendTo:"#codeApp",
+        revert: 'invalid',
+        zIndex: 2,
+        start: function() {
+          studioApp.resetButtonClick();
+        }
+      });
+      var scale = vizAppWidth / Applab.appWidth;
+      var gridSize = 20;
+      $('#visualization').droppable({
+        accept: '.new-design-element',
+        drop: function (event, ui) {
+          var elementType = ui.draggable[0].dataset.elementType;
+
+          var left = ui.position.left / scale;
+          left = Math.round(left - left % gridSize);
+          var top = ui.position.top / scale;
+          top = Math.round(top - top % gridSize);
+
+          Applab.createElement(elementType, left, top);
+        }
+      });
+    }
+
   }
 
   user = {applabUserId: config.applabUserId};
 };
 
 /**
+ * The types of acceptable HTML elements in the levelHtml.
+ * @type {{BUTTON: string, LABEL: string, INPUT: string}}
+ */
+var ElementType = {
+  BUTTON: 'button',
+  LABEL: 'label',
+  INPUT: 'input'
+};
+Applab.ElementType = ElementType;
+
+/**
+ * Returns an element id with the given prefix which is unused within
+ * the levelHtml.
+ * @param {string} prefix
+ * @returns {string}
+ */
+Applab.getUnusedElementId = function (prefix) {
+  var divApplab = $('#divApplab');
+  for (var i = 1; i < 100; i++) {
```

Comment:
Right now if I add 102 elements, the last two will be prefix.

What if we track nextId and use this as our starting point in the loop (instead of starting from 1 every time)? The only way a while loop wouldnt end here is if we had an infinite number of ids, right?